File: 2021/05/202105.py
import numpy as np


def parse(s):
    return np.loadtxt(s.replace(" -> ", ",").splitlines(),
                      delimiter=",", dtype=np.int32)


td = parse(open("tinput.txt").read())
data = parse(open("input.txt").read())


def part1(data, diag=False):
    x1, y1, x2, y2 = data[:, 0], data[:, 1], data[:, 2], data[:, 3]
    mx_x = max(x1.max(), x2.max())
    mx_y = max(y1.max(), y2.max())
    a = np.zeros((mx_y+1, mx_x+1), dtype=np.int32)
    for (x1, y1, x2, y2) in data:
        dx = -1 if x1 > x2 else 1
        dy = -1 if y1 > y2 else 1
        if y1 == y2:
            a[y1, x1:x2+dx:dx] += 1
        elif x1 == x2:
            a[y1:y2+dy:dy, x1] += 1
        else:
            if diag:  # diagonal included
                n = dx * (x2 - x1) + 1
                i = np.arange(n)
                a[y1 + i * dy, x1 + i * dx] += 1
    return (a > 1).sum()


assert part1(td) == 5
print("PART1:", part1(data))


def part2(data):
    return part1(data, diag=True)


assert part2(td) == 12
print("PART2:", part2(data))

File: 2021/05/202105.exs
defmodule M do
  def point(s) do
    s
    |> String.split(",")
    |> Enum.map(&String.to_integer/1)
  end

  def parse_line(s) do
    [l, r] = String.split(s, " -> ")
    [x1, y1] = point(l)
    [x2, y2] = point(r)
    {x1, y1, x2, y2}
  end

  def parse(input) do
    input
    |> String.split("\n")
    |> Enum.map(&parse_line/1)
  end

  def part1(lines) do
    lines
    |> Enum.reduce(%{}, fn {x1, y1, x2, y2}, m ->
      cond do
        x1 == x2 ->
          Enum.reduce(y1..y2, m, fn y, m ->
            Map.update(m, {x1, y}, 1, &(&1 + 1))
          end)

        y1 == y2 ->
          Enum.reduce(x1..x2, m, fn x, m ->
            Map.update(m, {x, y1}, 1, &(&1 + 1))
          end)

        true ->
          m
      end
    end)
    |> Map.values()
    |> Enum.count(fn v -> v > 1 end)
  end
end

IO.inspect(M.parse_line("830,945 -> 830,210"))
tinput = File.read!("tinput.txt")
tdata = M.parse(tinput)
IO.inspect(M.part1(tdata))

input = File.read!("input.txt")
data = M.parse(input)
IO.puts("PART1: #{M.part1(data)}")
